File: test-suite/src/test/java/io/micronaut/security/audit/docs/customconverter/AuthenticationToStringConverter.java
package io.micronaut.security.audit.docs.customconverter;

//tag::clazz[]
import io.micronaut.context.annotation.Replaces;
import io.micronaut.context.annotation.Requires;
import io.micronaut.core.convert.ConversionContext;
import io.micronaut.core.convert.TypeConverter;
import io.micronaut.security.converters.PrincipalToStringConverter;
import io.micronaut.security.authentication.Authentication;
import jakarta.inject.Singleton;

import java.util.Optional;

//end::clazz[]
@Requires(property = "spec.name", value = "CustomPrincipalConverterTest")
//tag::clazz[]
@Singleton
public class AuthenticationToStringConverter implements TypeConverter<Authentication, String> { // <1>
    @Override
    public Optional<String> convert(Authentication authentication, Class<String> targetType, ConversionContext context) {
        return Optional.ofNullable(authentication.getAttributes().get("CUSTOM_ID_ATTR")).map(Object::toString); // <2>
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/security/audit/docs/customconverter/AuthenticationToStringConverter.py
# tag::clazz[]
from jakarta.inject import Singleton
from java.util import Optional
from micronaut.context.annotation import Requires
from micronaut.core.convert import ConversionContext, TypeConverter
from micronaut.security.authentication import Authentication

# end::clazz[]
@Requires(property="spec.name", value="CustomPrincipalConverterTest")
# tag::clazz[]
@Singleton
class AuthenticationToStringConverter(TypeConverter[Authentication, str]):  # <1>
    def convert(self, authentication: Authentication, targetType: type[str], context: ConversionContext) -> Optional[str]:
        return Optional.ofNullable(authentication.getAttributes().get("CUSTOM_ID_ATTR")).map(lambda value: str(value))  # <2>
# end::clazz[]
